python/ucxx/ucxx/_lib_async/continuous_ucx_progress.py

```diff
@@ -24,12 +29,19 @@ def __init__(self, worker, event_loop):
         self.event_loop = event_loop
         self.asyncio_task = None
 
-    def __del__(self):
-        if self.asyncio_task is not None:
-            # FIXME: This does not work, the cancellation must be awaited.
-            # Running with polling mode will always cause
-            # `Task was destroyed but it is pending!` errors at ucxx.reset().
-            self.asyncio_task.cancel()
+        event_loop_close = self.event_loop.close
+
+        def _event_loop_close(*args, **kwargs):
+            if not self.event_loop.is_closed() and self.asyncio_task is not None:
+                try:
+                    self.asyncio_task.cancel()
+                    self.event_loop.run_until_complete(self.asyncio_task)
+                except asyncio.exceptions.CancelledError:
+                    pass
+                finally:
+                    event_loop_close(*args, **kwargs)
+
+        self.event_loop.close = _event_loop_close
```

Comment:
@wence- would you mind having one more look at this? It is a _real solution_ for the years long `coroutine was never awaited`/`Task was destroyed but it is pending!` warning that we've attempted to resolve in many instances, including https://github.com/rapidsai/ucx-py/pull/929, yet it is very intrusive. To me it doesn't look like it can be too harmful, but maybe you'll have some other thoughts or opinions.

Reply:
Some questions:

- This changes the behaviour of `EventLoop.close`. Does it only do so for this instance?
- What happens if multiple of these `ProgressTask`s are stacked up, I guess each one overwrites the `close` method, but remembers the previous one, so we do unwind everything?

Reply:
> * This changes the behaviour of `EventLoop.close`. Does it only do so for this instance?

That's right, the change only applies to `self.event_loop`, not the whole class, here's an example:

```python
import asyncio


async def run_patch():
    loop = asyncio.get_running_loop()
    print(f"run_patch: {loop}")

    loop_close = loop.close

    def _patch_close(*args, **kwargs):
        if not loop.is_closed():
            print("_patch_close")
            loop_close(*args, **kwargs)

    loop.close = _patch_close


async def run_orig():
    loop = asyncio.get_running_loop()
    print(f"run_orig: {loop}")


loop = asyncio.new_event_loop()
loop.run_until_complete(run_patch())
loop.close()

loop2 = asyncio.new_event_loop()
loop2.run_until_complete(run_orig())
loop2.close()
```

Which prints:

```
run_patch: <_UnixSelectorEventLoop running=True closed=False debug=False>
_patch_close
run_orig: <_UnixSelectorEventLoop running=True closed=False debug=False>
```

IOW, `_patch_close` only applies to `loop`, as expected.

> * What happens if multiple of these `ProgressTask`s are stacked up, I guess each one overwrites the `close` method, but remembers the previous one, so we do unwind everything?

This is a very good catch. Indeed this may not work and cause infinite recursion due to local `event_loop_close` in its original form, but it does work when passing the original `loop.close` function to a `partial`, such as:

```python
import asyncio
from functools import partial


async def run_patch():
    loop = asyncio.get_running_loop()
    print(f"run_patch: {loop}")

    loop_close = loop.close

    def _patch_close(loop_close, *args, **kwargs):
        if not loop.is_closed():
            print(f"_patch_close: {loop_close}")
            loop_close(*args, **kwargs)

    loop.close = partial(_patch_close, loop_close)

    loop_close = loop.close

    def _patch_close2(loop_close, *args, **kwargs):
        if not loop.is_closed():
            print(f"_patch_close2: {loop_close}")
            loop_close(*args, **kwargs)

    loop.close = partial(_patch_close2, loop_close)


loop = asyncio.new_event_loop()
loop.run_until_complete(run_patch())
loop.close()
```

If I'm not overlooking anything, the sample above is equivalent to having multiple `ProgressTask`s stacking up, so the unwinding occurs by always rewriting `loop.close` with the most recent `ProgressTask` that calls `loop.close` that was previously set. This change is now reflected in https://github.com/rapidsai/ucxx/pull/116/commits/c5c2ceb0514f0bf652a4ffe48036befcba6bfc66 .